### PR TITLE
merging pulse implementation with main sketch

### PR DESCRIPTION
**Pull request by spartaninzaghi**
Included code for pulse functionality

### CODE REVIEW

File: arduino/arduino.ino [hunk 1]
```diff
@@ -6,6 +10,74 @@ enum CommandMode {
...
+  volatile uint8_t portRegister = *info.port;
+
+  if (state) {
+    portRegister |= 1 << info.bit;
+  } else {
+    portRegister &= ~(1 << info.bit);
+  }
+  
+  *info.port = portRegister;
+}
```
Comment (tufourn) [inline arduino/arduino.ino:54]: change writePin to something like digitalWriteFast().
create const array of PinInfo for all pins on uno, not just nozzle pins.
then we can also utilize digitalWriteFast() for other stuff like stepper controls

File: arduino/arduino.ino [hunk 2]
```diff
@@ -27,10 +99,22 @@ void gotoNextLine() {
 void dispense(unsigned char mask) {
   if (cmdMode == MASK1) {
     /* TODO: dispense first set of nozzles */
```
Comment (tufourn) [inline arduino/arduino.ino:101]: delete todo comment

File: arduino/arduino.ino [hunk 3]
```diff
@@ -27,10 +99,22 @@ void gotoNextLine() {
...
+      // pulse nozzle at index `i` if its bit is set HIGH in mask
+      if (mask & (1 << i)){
+        pulse(nozzlePins[i]);
+      }
+    }
     cmdMode = MASK2;
   }
   if (cmdMode == MASK2) {
     /* TODO: dispense second set of nozzles */
+    for (int i = 8; i < 12; i++){
```
Comment (tufourn) [inline arduino/arduino.ino:112]: wrong code, you can't shift by more than 8

File: arduino/arduino.ino [hunk 4]
```diff
@@ -6,6 +10,74 @@ enum CommandMode {
...
+  volatile uint8_t* pin;
+  volatile uint8_t* port;
+  uint8_t bit;
+};
+
+// Nozzle Number       1  2  3  4  5  6  7  8  9   10  11  12 
+const int nozzles[] = {2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 15};
+
+const PinInfo nozzlePins[] = {
+//                           Register  Pin Nozzle
```
Comment (tufourn) [inline arduino/arduino.ino:24]: D2, D3, etc are not registers, just gpio pins on the atmega chip.
don't need the comment for nozzle number here.

File: arduino/arduino.ino [hunk 5]
```diff
@@ -6,6 +10,74 @@ enum CommandMode {
...
+  writePin(info, HIGH); 
+  delayMicroseconds(PULSE_DURATION);
+  writePin(info, LOW);
+}
+
+/**
+ * Test all nozzle pins successively
+ */
+void pulseTestSuccessive(){
+  for (int i = 0; i < 12; i++){
```
Comment (tufourn) [inline arduino/arduino.ino:74]: explain why you're doing these delays here, and add a link to the inkshield theory page

File: arduino/arduino.ino [hunk 6]
```diff
@@ -6,6 +10,74 @@ enum CommandMode {
...
+    portRegister &= ~(1 << info.bit);
+  }
+  
+  *info.port = portRegister;
+}
+
+/**
+ * Initiate a 5V pulse on a single nozzle
+ *
+ * For HP6602, A pulse lasts for 5us, after which follows a delay.
```
Comment (tufourn) [inline arduino/arduino.ino:59]: maybe move this doc to the pulseTestSuccessive() function. doc of pulse() should just be something like "pulse specified pin by PULSE_DURATION us"

File: arduino/arduino.ino (hunk 1 above)
Comment (spartaninzaghi) [reply arduino/arduino.ino:54]: I created the const array for all pins and generalized the digitalWriteFast() function. I currently have some redundancy in having both the allPins const array, and the nozzlePins array. I can refactor this once we have some surety on the pin assignments for other components like stepper motors, and also when we finalize the microcontroller to be used (uno vs. mega). If we do not end up using all 12 nozzles with the uno, I could get rid of the redundancy much easier. I could also use a for loop to assign the 1st 11 nozzlePins contiguously, and the 12th in a separate assignment.

File: arduino/arduino.ino (hunk 2 above)
Comment (spartaninzaghi) [reply arduino/arduino.ino:101]: Done

File: arduino/arduino.ino (hunk 3 above)
Comment (spartaninzaghi) [reply arduino/arduino.ino:112]: Fixed

File: arduino/arduino.ino [hunk 7]
```diff
@@ -26,11 +139,21 @@ void gotoNextLine() {
 // when last mask done, call step() and change cmdMode to INSTRUCTION
```
[the remark this answers is not in the document]
Comment (spartaninzaghi) [reply arduino/arduino.ino:139]: That’s a very key consideration. I’ll make room and factor that in (adaptability to different nozzle counts and error checking)

**Pull request merged**